src/stactools/core/utils/raster_footprint.py

```diff
@@ -0,0 +1,286 @@
+import logging
+from itertools import groupby
+from typing import Any, Dict, Iterator, List, Optional, Tuple
+
+import numpy
+import numpy as np
+import rasterio
+import rasterio.features
+from pystac import Item
+from rasterio.crs import CRS
+from rasterio.warp import transform_geom
+from shapely.geometry import mapping, shape
+from shapely.geometry.multipolygon import MultiPolygon
+from shapely.geometry.polygon import Polygon
+
+logger = logging.getLogger(__name__)
+
+# Roughly 1 centimeter in geodetic coordinates
+DEFAULT_PRECISION = 7
+
+
+def update_geometry_from_asset_footprint(
+    item: Item,
+    *,
+    asset_names: List[str] = [],
+    precision: int = DEFAULT_PRECISION,
+    densification_factor: Optional[int] = None,
+    simplify_tolerance: Optional[float] = None,
+    no_data: Optional[int] = None,
+) -> bool:
+    """
+    Accepts an Item and an optional list of asset names within that item, and updates
+    the geometry of that Item in-place with the data footprint derived from the first
+    of the assets that exists in the Item.
+
+    Two important operations during this calculation are the densification of the
+    footprint in the native CRS and simplification of the footprint after reprojection.
+    If the initial low-vertex polygon in the native CRS is not densified, this can result
+    in a reprojected polygon that does not accurately represent the data footprint. For
+    example, a MODIS asset represented by a rectangular 5 point Polygon in a sinusoidal
+    projection will reproject to a parallelogram in EPSG 4326, when it would be more
+    accurately represented by a polygon with two parallel sides and two curved sides.
+    The difference between these representations is even greater the further away from
+    the meridian and equator the asset is located.
+
+    After reprojection to EPSG 4326, the footprint may have more points than desired.
+    This can be simplified to a polygon with fewer points that maintain a minimum distance
+    to the original geometry.
+
+    Args:
+        item (Item): The PySTAC Item to extend.
+        asset_names (List[str]): The names of the assets for which to extract footprints. If the
```

Comment:
This, to me, is not clear. It implies that footprints will be extracted for all assets in the list. Which then is not clear how the multiple footprints from the multiple assets would be combined. But this is not the case; IIUC, the first successful footprint from the list of assets is used - all remaining assets are ignored. Likewise, if no asset list is provided, then the footprint generated from first successful asset in the Item is used. Is that right?

Reply:
> the first successful footprint from the list of assets is used - all remaining assets are ignored

Correct. Which brings up the interesting question on whether asset order is preserved when loading into PySTAC. If not, results could be non-deterministic.

> if no asset list is provided, then the footprint generated from first successful asset in the Item is used. Is that right?

Correct.

Reply:
I'll revise the text to be more clear.